modules/Bio/EnsEMBL/Variation/DBSQL/PublicationAdaptor.pm

```diff
@@ -206,22 +206,51 @@ sub fetch_all_by_dbID_list {
 sub fetch_all_by_Variation {
     my $self    = shift;
     my $var_obj = shift;
-    
+
     if(!defined($var_obj) || ref($var_obj ne 'Variation')) {
-        throw("variation argument is required");
+        throw("Variation argument is required");
     }
-    
+
+    my $variation_id = $var_obj->dbID;
+
     my @pub;
     my $publication_id;
+    my $data_source_attrib;
+    my $attrib_id_to_value = {};
 
-    my $sth = $self->prepare(qq{SELECT  publication_id from variation_citation where variation_id = ?  });
-    $sth->execute($var_obj->dbID);
-    $sth->bind_columns(\$publication_id);
+    my $sth = $self->prepare(qq{SELECT  publication_id, data_source_attrib from variation_citation where variation_id = ?  });
+    $sth->execute($variation_id);
+    $sth->bind_columns(\$publication_id, \$data_source_attrib);
     while ($sth->fetch()){
-        push @pub, $self->fetch_by_dbID($publication_id)
+
+      my $publication = $self->fetch_by_dbID($publication_id);
+
+      if($data_source_attrib) {
+        my @sources = split /,/, $data_source_attrib;
+        foreach my $source_attrib (@sources) {
+          my $source = $attrib_id_to_value->{$source_attrib};
+
+          if(!defined($source)) {
+            my $sth_attrib = $self->prepare(qq{SELECT value FROM attrib WHERE attrib_id = ? });
```

Comment:
You can use the AttributeAdaptor for getting the value with: attrib_value_for_id(). We use it in the the PhenotypeAdaptor: $self->db->get_AttributeAdaptor->attrib_value_for_id($id). I would suggest to first initialise the adaptor my $attribute_adaptor = $self->db->get_AttributeAdaptor; And then you can use it whenever you need to look up a value: $attribute_adaptor->attrib_value_for_id($data_source_attrib).